### PR TITLE
Add Sentinel support (PEP-661)

### LINKED ISSUE #3207: Support Sentinel types in pydantic models
### Describe the Bug

```
import pydantic
from pydantic.experimental.missing_sentinel import MISSING

class UpdateUserPayload(pydantic.BaseModel):
    phone_number: str | MISSING = MISSING 
    # Errors with "Expected a type form, got instance of `type[str] | UnionType`". Should not error.
```

This is an experimental feature, but required if you have a `PATCH` model and the generated schema shouldn't erroneously indicate that the value is nullable.

https://pydantic.dev/docs/validation/latest/concepts/experimental/#missing-sentinel

### Sandbox Link

_No response_

### (Only applicable for extension issues) IDE Information

_No response_

### PR DESCRIPTION
**Pull request by hrolfurgylfa**
# Summary

This PR adds support for PEP 661 Sentinel types.

Fixes #3207

# Test Plan

I've run this as an LSP for a bit on a bigger private repo and it seems to work, but the main testing was the tests added to `test/narrow.rs` and `test/sentinel.rs`.

# Some assumptions/things I didn't fully understand and might deserve more attention:

- I'm using NestingContext::toplevel() when creating a Sentinel struct, this is similar to TypeVar/TypeAlias/a few others, I left the same TODO in place as they had. I looked through some of the other types and couldn't find any other way it was being done, so I left it like this for now, but if anyone has any pointers to how it should be done better, please let me know and I can fix it.

- For the cinderx support, I added the sentinel to it in the TODO section of variables that get set to Any. I haven't really diven into what cinderx is and how the support for it is implemented, but if it is important to get it working I can go ahead and look into that.

- I'm using bind_legacy_type_var_or_typing_alias to bind the Sentinel type, I think this should be ok as the PEP is pretty strict in how it can be assigned. I've done some testing and only found one edge case so far that fails because of this, but I don't think Final's should be mixed with Sentinel objects like this anyway, here is the test case that fails:

```py
from typing import Any, Final
from typing_extensions import Sentinel

MISSING: Final[Any]
MISSING = Sentinel("MISSING")
```

It fails because the `Final` doesn't notice the assignment, it still thinks the variable was never assigned to.

# Status

This PR is mostly ready for review, but I'm keeping it in Draft form until #3562 is addressed, since until then I can't finalize Python 3.15 support.

### CODE REVIEW

**Comment by yangdanny97** [comment]
Thanks for the PR, wow! that was fast, the PEP just got accepted recently and the conformance tests haven't even been added yet

I see you added lots of tests already, but maybe worth looking at the proposed conformance test to see if there's anything missing
https://github.com/python/typing/pull/2277/changes

I can give this a more detailed review once I've unblocked you on https://github.com/facebook/pyrefly/issues/3562

**Comment by hrolfurgylfa** [comment]
Thanks, I didn't know those conformance tests existed, I'll review those once I get back home after the weekend, don't have my laptop with me.

And yeah, a proper review should probably be held until I can add the Python 3.15 support, since there is a bit of extra code to differentiate sentinel lowercase builtin and Sentinel uppercase from typing extensions, might be adding another 30-50 lines, don't remember the exact number though.

I'm hoping to have both of those ready on Tuesday/Wednesday evening (UTC+0).

**Comment by yangdanny97** [comment]
No rush at all. The typeshed update just got merged, so once you rebase it should be there

**Comment by hrolfurgylfa** [comment]
This PR should now be ready for review.

It now passes the [conformance tests](https://github.com/python/typing/pull/2277/changes), the only change was just no longer emit an error when a sentinel is assigned to a variable with a different name like `MISSING = sentinel("<MISSING>")`. This is something the PEP prohibited but in the [open conformance PR to python/typing](https://github.com/python/typing/pull/2277) that prohibition was lifted, so type checkers are no longer supposed to emit an error there.

I'm still not perfectly sure on the query interface and tsp implementation for sentinel, I haven't really figured out how to test the tsp stuff properly and while I have looked at the output of the shape query with sentinels I'm not perfectly sure how it should look, as such I haven't added any automated tests for it.

File: crates/pyrefly_types/src/stdlib.rs [hunk 1]
```diff
@@ -257,6 +263,9 @@ impl Stdlib {
             param_spec_kwargs: lookup_concrete(standardised(3, 10), "ParamSpecKwargs"),
             type_var_tuple: lookup_concrete(standardised(3, 11), "TypeVarTuple"),
             type_alias_type: lookup_concrete(standardised(3, 12), "TypeAliasType"),
+            // sentinel: lookup_concrete(typing_extensions, "Sentinel"),
+            sentinel_builtin: lookup_concrete(builtins, "sentinel"),
```
Comment (rchen152) [inline crates/pyrefly_types/src/stdlib.rs]: How different are `builtins.sentinel` and `typing_extensions.Sentinel`? If they're pretty similar, I'd suggest combining them into just one `sentinel`, which is looked up from builtins in 3.15+ and from typing_extensions otherwise, and modifying the rest of the PR to likewise stop tracking the distinction between the two.

In general, we treat the `typing` and `typing_extensions` versions of things as the same without worrying about the minor differences between their runtime definitions. And distinguishing between `builtins.sentinel` and `typing_extensions.Sentinel` seems to be adding a fair bit of complexity to this PR.

File: crates/pyrefly_types/src/sentinel.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+struct SentinelInner {
+    qname: QName,
+    kind: SentinelKind,
+}
+
+impl Sentinel {
+    pub fn new(name: Identifier, module: Module, kind: SentinelKind) -> Self {
+        Self(ArcId::new(SentinelInner {
+            kind,
+            // TODO: properly take parent from caller of new()
```
Comment (rchen152) [inline crates/pyrefly_types/src/sentinel.rs]: You can grab the nesting context from the Scopes stack when constructing Binding::Sentinel - see comment in stmt.rs.

File: pyrefly/lib/binding/stmt.rs [hunk 3]
```diff
@@ -281,6 +282,36 @@ impl<'a> BindingsBuilder<'a> {
...
+            self.ensure_expr(expr, static_type_usage);
+        }
+        for kw in call.arguments.keywords.iter_mut() {
+            self.ensure_expr(&mut kw.value, static_type_usage);
+        }
+        // Like legacy type var, Sentinel can only be created with a single Sentinel binding to a
+        // single variable (https://peps.python.org/pep-0661/#typing). Thus we bind it in the same
+        // way legacy type vars are bound.
+        self.bind_legacy_type_var_or_typing_alias(name, |ann| {
+            Binding::Sentinel(Box::new((
```
Comment (rchen152) [inline pyrefly/lib/binding/stmt.rs:301]: Here you should have access to `self.scopes`, from which you can get the current nesting context: https://github.com/facebook/pyrefly/blob/4d820b2eb582b4c50de23f6e2d9998427d59e6a0/pyrefly/lib/binding/scope.rs#L1513.

File: website/docs/error-kinds.mdx [hunk 4]
```diff
@@ -885,6 +885,20 @@ class TD(TypedDict):
...
 ```
 
+## invalid-sentinel
+
+An error caused by incorrect definition of a Sentinel. A few examples:
+
+```python
+from typing_extensions import Sentinel
+
+# Sentinels must be assigned to a matching variable.
```
Comment (rchen152) [inline website/docs/error-kinds.mdx]: This doc example looks incorrect; the implementation (correctly) allows the name to not match the variable name.

**Review by rchen152 — CHANGES_REQUESTED** [review]
Thank you for the PR! This is great. I left a few comments, mostly minor. The one larger question I have is whether we can drop the distinction between `builtins.sentinel` and `typing_extensions.Sentinel`, which seems like it would simplify the code a fair bit if we can get away with it.

File: crates/pyrefly_types/src/stdlib.rs (hunk 1 above)
Comment (hrolfurgylfa) [reply crates/pyrefly_types/src/stdlib.rs]: The main reason I differentiated between the two was the `__name__` attribute that is only available on the builtins sentinel, not the typing extensions Sentinel. If we merge them it would mean that in Python 3.15+ pyrefly would say that accessing `__name__` on typing_extensions Sentinel is allowed and returns a string, while raising a runtime error if run. If that is fine I can go ahead and remove this logic.

Here is how I confirmed this difference if you want to test it with the python 3.15 beta:
```bash
uv run -p 3.15 --with 'typing_extensions' python -c 'from typing_extensions import Sentinel; MISSING = Sentinel("MISSING"); print(MISSING.__name__)'
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>     from typing_extensions import Sentinel; MISSING = Sentinel("MISSING"); MISSING.__name__
>                                                                            ^^^^^^^^^^^^^^^^
> AttributeError: 'Sentinel' object has no attribute '__name__'. Did you mean '.__ne__' instead of '.__name__'?

uv run -p 3.15 --with 'typing_extensions' python -c 'MISSING = sentinel("MISSING"); print(MISSING.__name__)'
> MISSING
```

Also, if this ends up being kept, I should probably add a comment explaining the reasoning better, and remove that commented out line, not sure how that one got past me.

File: crates/pyrefly_types/src/sentinel.rs (hunk 2 above)
Comment (hrolfurgylfa) [reply crates/pyrefly_types/src/sentinel.rs]: Thanks, fixed

File: website/docs/error-kinds.mdx (hunk 4 above)
Comment (hrolfurgylfa) [reply website/docs/error-kinds.mdx]: Good catch, forgot to change this when I removed the error on different variable name, fixed

File: crates/pyrefly_types/src/stdlib.rs (hunk 1 above)
Comment (rchen152) [reply crates/pyrefly_types/src/stdlib.rs]: Thanks for the explanation! If the presence/absence of the `__name__` attribute is the main difference, then I'm in favor of merging them.

**Review by rchen152 — APPROVED** [review]
Looks good, thank you! I'll get a second review and get this merged.

**Comment by hrolfurgylfa** [comment]
Thanks :smile:

**Review by yangdanny97 — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.

**Comment by yangdanny97** [comment]
Great work on this PR @hrolfurgylfa, thanks!
